File: packages/js/node/src/swap/fixedfloat-transport.ts
/**
 * Everything that crosses the wire to a FixedFloat-compatible API: the
 * HMAC-signed `/api/v2` POST, envelope parsing into data or FixedFloatApiError,
 * request/response log surfacing, and weight-budget reserve / 429 accounting.
 * The unauthenticated XML rates GET lives in fixedfloat-rates.ts.
 */

import { createHmac } from "node:crypto";
import { optionalCoercedString } from "./fixedfloat-fields.ts";
import type { SwapProviderApiRequestLog, SwapProviderApiResponseLog } from "./provider.ts";

/** The process-local request weight guard the service attaches (see weight-budget.ts). */
export interface FixedFloatWeightBudget {
  reserve(path: string): Promise<void>;
  markRateLimited(): Promise<void>;
  canReserve(path: string): Promise<boolean>;
}

interface FixedFloatEnvelope {
  readonly code?: unknown;
  readonly msg?: unknown;
  readonly data?: unknown;
}

export class FixedFloatApiError extends Error {
  readonly path: string;
  readonly kind: "api" | "http" | "invalid_json" | "network" | "rate_limited" | "timeout";
  readonly status?: number;
  readonly fixedFloatCode?: unknown;
  readonly fixedFloatMessage?: string;

  constructor(input: {
    readonly path: string;
    readonly kind: FixedFloatApiError["kind"];
    readonly status?: number;
    readonly fixedFloatCode?: unknown;
    readonly fixedFloatMessage?: string;
    readonly message: string;
    readonly cause?: unknown;
  }) {
    super(input.message, input.cause === undefined ? undefined : { cause: input.cause });
    this.name = "FixedFloatApiError";
    this.path = input.path;
    this.kind = input.kind;
    this.status = input.status;
    this.fixedFloatCode = input.fixedFloatCode;
    this.fixedFloatMessage = input.fixedFloatMessage;
  }

  static fromFetchError(path: string, error: unknown): FixedFloatApiError {
    const aborted =
      error instanceof Error &&
      (error.name === "AbortError" || error.message.toLowerCase().includes("abort"));
    return new FixedFloatApiError({
      path,
      kind: aborted ? "timeout" : "network",
      message: aborted
        ? `FixedFloat ${path} request timed out.`
        : `FixedFloat ${path} request failed before a response was received.`,
      cause: error,
    });
  }
}

export class FixedFloatTransport {
  /** Already stripped of trailing slashes; also the origin of the XML rates GET. */
  readonly baseUrl: string;
  readonly fetcher: typeof globalThis.fetch;
  readonly requestTimeoutMs: number;
  private readonly provider: string;
  private readonly key: string;
  private readonly secret: string;
  private apiRequestLogger: ((entry: SwapProviderApiRequestLog) => void) | undefined;
  private apiResponseLogger: ((entry: SwapProviderApiResponseLog) => void) | undefined;
  private weightBudget: FixedFloatWeightBudget | undefined;

  constructor(input: {
    /** Provider name stamped on every log entry. */
    readonly provider: string;
    readonly key: string;
    readonly secret: string;
    readonly baseUrl: string;
    readonly fetch: typeof globalThis.fetch;
    readonly requestTimeoutMs: number;
  }) {
    this.provider = input.provider;
    this.key = input.key;
    this.secret = input.secret;
    this.baseUrl = input.baseUrl;
    this.fetcher = input.fetch;
    this.requestTimeoutMs = input.requestTimeoutMs;
  }

  attachApiRequestLogger(log: (entry: SwapProviderApiRequestLog) => void): void {
    this.apiRequestLogger = log;
  }

  attachApiResponseLogger(log: (entry: SwapProviderApiResponseLog) => void): void {
    this.apiResponseLogger = log;
  }

  attachWeightBudget(budget: FixedFloatWeightBudget): void {
    this.weightBudget = budget;
  }

  async canAcceptRequest(path: string): Promise<boolean> {
    if (this.weightBudget === undefined) return true;
    return await this.weightBudget.canReserve(path);
  }

  async post(path: string, body: Record<string, unknown>): Promise<unknown> {
    if (this.weightBudget !== undefined) {
      await this.weightBudget.reserve(path);
    }
    const bodyString = JSON.stringify(body);
    // Surface every outbound request before the call. The service sink sanitizes
    // nested secrets (e.g. the order token on status/refund bodies); the API key
    // and HMAC signature live in headers and are deliberately never logged.
    this.logApiRequest(path, body);
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.requestTimeoutMs);
    let response: Response;
    try {
      response = await this.fetcher(`${this.baseUrl}/api/v2/${path}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json; charset=UTF-8",
          "X-API-KEY": this.key,
          "X-API-SIGN": createHmac("sha256", this.secret).update(bodyString).digest("hex"),
        },
        body: bodyString,
        signal: controller.signal,
      });
    } catch (error) {
      const apiError = FixedFloatApiError.fromFetchError(path, error);
      this.logApiResponse({
        path,
        status: 0,
        ok: false,
        msg: apiError.message,
      });
      throw apiError;
    } finally {
      clearTimeout(timeout);
    }
    const text = await response.text();
    let parsed: FixedFloatEnvelope;
    try {
      parsed = text.length === 0 ? {} : (JSON.parse(text) as FixedFloatEnvelope);
    } catch (error) {
      this.logApiResponse({
        path,
        status: response.status,
        ok: false,
        msg: `FixedFloat ${path} returned invalid JSON.`,
      });
      throw new FixedFloatApiError({
        path,
        kind: "invalid_json",
        status: response.status,
        message: `FixedFloat ${path} returned invalid JSON.`,
        cause: error,
      });
    }
    // Surface every response (including API-error envelopes) before any throw. The
    // service sink sanitizes nested secrets — notably the order token in a
    // create/order response — so this must not pre-redact.
    this.logApiResponse({
      path,
      status: response.status,
      ok: response.ok,
      code: parsed.code,
      msg: parsed.msg,
      data: parsed.data,
    });
    if (!response.ok) {
      if (response.status === 429 && this.weightBudget !== undefined) {
        await this.weightBudget.markRateLimited();
      }
      throw new FixedFloatApiError({
        path,
        kind: response.status === 429 ? "rate_limited" : "http",
        status: response.status,
        fixedFloatMessage: optionalCoercedString(parsed.msg),
        message: formatFixedFloatApiErrorMessage(path, response.status, parsed.msg),
      });
    }
    if (parsed.code !== 0) {
      throw new FixedFloatApiError({
        path,
        kind: "api",
        fixedFloatCode: parsed.code,
        fixedFloatMessage: optionalCoercedString(parsed.msg),
        message: typeof parsed.msg === "string" ? parsed.msg : `FixedFloat ${path} failed.`,
      });
    }
    return parsed.data;
  }

  logApiRequest(path: string, body: Record<string, unknown> = {}): void {
    this.apiRequestLogger?.({
      provider: this.provider,
      path,
      body,
    });
  }

  logApiResponse(input: {
    readonly path: string;
    readonly status: number;
    readonly ok: boolean;
    readonly code?: unknown;
    readonly msg?: unknown;
    readonly data?: unknown;
  }): void {
    this.apiResponseLogger?.({
      provider: this.provider,
      path: input.path,
      status: input.status,
      ok: input.ok,
      code: input.code,
      msg: input.msg,
      data: input.data,
    });
  }
}

function formatFixedFloatApiErrorMessage(path: string, status: number, msg: unknown): string {
  const fixedFloatMessage = optionalCoercedString(msg);
  return fixedFloatMessage === undefined
    ? `FixedFloat ${path} failed with HTTP ${status}.`
    : `FixedFloat ${path} failed with HTTP ${status}: ${fixedFloatMessage}`;
}

File: packages/dotnet/BTCPayServer.Plugins.OpenReceive/Swaps/FixedFloatTransport.cs
#nullable enable
using System;
using System.Globalization;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;

namespace BTCPayServer.Plugins.OpenReceive.Swaps;

public enum FixedFloatApiErrorKind
{
    /// <summary>The envelope answered with a non-zero <c>code</c>.</summary>
    Api,
    /// <summary>A non-2xx HTTP status other than 429.</summary>
    Http,
    /// <summary>The body was not JSON.</summary>
    InvalidJson,
    /// <summary>No response was received.</summary>
    Network,
    /// <summary>HTTP 429.</summary>
    RateLimited,
    /// <summary>The request timed out.</summary>
    Timeout,
}

/// <summary>A FixedFloat-compatible API call that did not return usable data.</summary>
public sealed class FixedFloatApiException : Exception
{
    public string Path { get; }
    public FixedFloatApiErrorKind Kind { get; }
    public int? Status { get; }
    public string? FixedFloatCode { get; }
    public string? FixedFloatMessage { get; }

    public FixedFloatApiException(
        string path,
        FixedFloatApiErrorKind kind,
        string message,
        int? status = null,
        string? fixedFloatCode = null,
        string? fixedFloatMessage = null,
        Exception? cause = null)
        : base(message, cause)
    {
        Path = path;
        Kind = kind;
        Status = status;
        FixedFloatCode = fixedFloatCode;
        FixedFloatMessage = fixedFloatMessage;
    }
}

/// <summary>
/// Everything that crosses the wire to a FixedFloat-compatible API: the HMAC-signed
/// <c>/api/v2</c> POST, envelope parsing into data or <see cref="FixedFloatApiException"/>,
/// request/response log surfacing, weight-budget reserve / 429 accounting, and the
/// unauthenticated GET for the public XML rates export.
/// </summary>
public sealed class FixedFloatTransport
{
    private static readonly JsonSerializerOptions BodyJsonOptions = new()
    {
        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    /// <summary>Already stripped of trailing slashes; also the origin of the XML rates GET.</summary>
    public string BaseUrl { get; }
    public TimeSpan RequestTimeout { get; }

    private readonly HttpClient _httpClient;
    private readonly string _provider;
    private readonly string _key;
    private readonly byte[] _secret;
    private Action<SwapProviderApiRequestLog>? _apiRequestLogger;
    private Action<SwapProviderApiResponseLog>? _apiResponseLogger;
    private ISwapWeightBudget? _weightBudget;

    public FixedFloatTransport(HttpClient httpClient, string key, string secret, string baseUrl, string provider, TimeSpan requestTimeout)
    {
        _httpClient = httpClient;
        _key = key;
        _secret = Encoding.UTF8.GetBytes(secret);
        BaseUrl = baseUrl.TrimEnd('/');
        _provider = provider;
        RequestTimeout = requestTimeout;
    }

    public void AttachApiRequestLogger(Action<SwapProviderApiRequestLog> log) => _apiRequestLogger = log;
    public void AttachApiResponseLogger(Action<SwapProviderApiResponseLog> log) => _apiResponseLogger = log;
    public void AttachWeightBudget(ISwapWeightBudget budget) => _weightBudget = budget;

    /// <summary>POST <c>{baseUrl}/api/v2/{path}</c> and return the envelope's <c>data</c>.</summary>
    public async Task<JsonNode?> PostAsync(string path, JsonObject body, CancellationToken cancellationToken)
    {
        _weightBudget?.Reserve(path);
        var bodyString = body.ToJsonString(BodyJsonOptions);
        var bodyBytes = Encoding.UTF8.GetBytes(bodyString);
        // Surface every outbound request before the call. The API key and HMAC
        // signature live in headers and are deliberately never logged; the order
        // token (status/refund bodies) is redacted before the entry leaves here.
        LogApiRequest(path, RedactedBodyJson(body));

        HttpResponseMessage response;
        string text;
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{BaseUrl}/api/v2/{path}");
            request.Content = new ByteArrayContent(bodyBytes);
            request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json") { CharSet = "UTF-8" };
            request.Headers.TryAddWithoutValidation("X-API-KEY", _key);
            request.Headers.TryAddWithoutValidation("X-API-SIGN", Sign(bodyBytes));
            using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            timeout.CancelAfter(RequestTimeout);
            response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseContentRead, timeout.Token).ConfigureAwait(false);
            text = await response.Content.ReadAsStringAsync(timeout.Token).ConfigureAwait(false);
        }
        catch (Exception error) when (error is not OperationCanceledException || !cancellationToken.IsCancellationRequested)
        {
            var apiError = FromTransportError(path, error);
            LogApiResponse(path, 0, false, null, apiError.Message);
            throw apiError;
        }

        using (response)
        {
            var status = (int)response.StatusCode;
            JsonObject envelope;
            try
            {
                envelope = text.Length == 0 ? new JsonObject() : FixedFloatFields.ObjectOrEmpty(JsonNode.Parse(text));
            }
            catch (JsonException error)
            {
                var message = $"FixedFloat {path} returned invalid JSON.";
                LogApiResponse(path, status, false, null, message);
                throw new FixedFloatApiException(path, FixedFloatApiErrorKind.InvalidJson, message, status, cause: error);
            }
            var code = envelope["code"];
            var msg = envelope["msg"];
            var codeText = FixedFloatFields.OptionalCoercedString(code) ?? (code is null ? null : code.ToJsonString());
            var msgText = FixedFloatFields.OptionalCoercedString(msg);
            LogApiResponse(path, status, response.IsSuccessStatusCode, codeText, msgText);
            if (!response.IsSuccessStatusCode)
            {
                var rateLimited = response.StatusCode == HttpStatusCode.TooManyRequests;
                if (rateLimited) _weightBudget?.MarkRateLimited();
                throw new FixedFloatApiException(
                    path,
                    rateLimited ? FixedFloatApiErrorKind.RateLimited : FixedFloatApiErrorKind.Http,
                    msgText is null
                        ? $"FixedFloat {path} failed with HTTP {status}."
                        : $"FixedFloat {path} failed with HTTP {status}: {msgText}",
                    status,
                    fixedFloatMessage: msgText);
            }
            if (!IsZero(code))
            {
                throw new FixedFloatApiException(
                    path,
                    FixedFloatApiErrorKind.Api,
                    msg is JsonValue && msg.GetValueKind() == JsonValueKind.String ? msg.GetValue<string>() : $"FixedFloat {path} failed.",
                    fixedFloatCode: codeText,
                    fixedFloatMessage: msgText);
            }
            return envelope["data"];
        }
    }

    /// <summary>GET <c>{baseUrl}{absolutePath}</c> as text (the public XML rates export).</summary>
    public async Task<string> GetTextAsync(string absolutePath, CancellationToken cancellationToken)
    {
        var name = absolutePath[(absolutePath.LastIndexOf('/') + 1)..];
        HttpResponseMessage response;
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{BaseUrl}{absolutePath}");
            request.Headers.TryAddWithoutValidation("Accept", "application/xml, text/xml, */*");
            using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            timeout.CancelAfter(RequestTimeout);
            response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseContentRead, timeout.Token).ConfigureAwait(false);
            using (response)
            {
                if (!response.IsSuccessStatusCode)
                {
                    throw new InvalidOperationException($"FixedFloat rates {name} failed with HTTP {(int)response.StatusCode}.");
                }
                return await response.Content.ReadAsStringAsync(timeout.Token).ConfigureAwait(false);
            }
        }
        catch (Exception error) when (error is not InvalidOperationException &&
                                      (error is not OperationCanceledException || !cancellationToken.IsCancellationRequested))
        {
            var timedOut = IsTimeout(error);
            throw new InvalidOperationException(
                timedOut
                    ? $"FixedFloat rates {name} request timed out."
                    : $"FixedFloat rates {name} request failed before a response was received.",
                error);
        }
    }

    public void LogApiRequest(string path, string? bodyJson = null) =>
        _apiRequestLogger?.Invoke(new SwapProviderApiRequestLog(_provider, path, bodyJson));

    public void LogApiResponse(string path, int status, bool ok, string? code, string? message) =>
        _apiResponseLogger?.Invoke(new SwapProviderApiResponseLog(_provider, path, status, ok, code, message));

    private static string RedactedBodyJson(JsonObject body)
    {
        if (!body.ContainsKey("token")) return body.ToJsonString(BodyJsonOptions);
        var redacted = (JsonObject)body.DeepClone();
        redacted["token"] = "[redacted]";
        return redacted.ToJsonString(BodyJsonOptions);
    }

    private string Sign(byte[] bodyBytes) =>
        Convert.ToHexString(HMACSHA256.HashData(_secret, bodyBytes)).ToLowerInvariant();

    private static bool IsZero(JsonNode? code)
    {
        if (code is not JsonValue value) return false;
        if (value.TryGetValue<JsonElement>(out var element))
        {
            return element.ValueKind == JsonValueKind.Number && element.TryGetDouble(out var real) && real == 0;
        }
        if (value.TryGetValue<int>(out var i)) return i == 0;
        if (value.TryGetValue<long>(out var l)) return l == 0;
        if (value.TryGetValue<double>(out var d)) return d == 0;
        if (value.TryGetValue<decimal>(out var m)) return m == 0;
        return false;
    }

    private static bool IsTimeout(Exception error) =>
        error is OperationCanceledException ||
        error is TimeoutException ||
        error.InnerException is TimeoutException;

    private static FixedFloatApiException FromTransportError(string path, Exception error)
    {
        var timedOut = IsTimeout(error);
        return new FixedFloatApiException(
            path,
            timedOut ? FixedFloatApiErrorKind.Timeout : FixedFloatApiErrorKind.Network,
            timedOut
                ? $"FixedFloat {path} request timed out."
                : $"FixedFloat {path} request failed before a response was received.",
            cause: error);
    }
}
